bpf/protocol_http2.h

```diff
@@ -95,110 +125,219 @@ http2_grpc_end(http2_conn_stream_t *stream, http2_grpc_request_t *prev_info, voi
     bpf_map_delete_elem(&ongoing_http2_grpc, stream);
 }
 
-static __always_inline void process_http2_grpc_frames(pid_connection_info_t *pid_conn,
-                                                      void *u_buf,
-                                                      int bytes_len,
-                                                      u8 direction,
-                                                      u8 ssl,
-                                                      u16 orig_dport) {
-    int pos = 0;
-    u8 found_start_frame = 0;
-    u8 found_end_frame = 0;
-
-    http2_grpc_request_t *prev_info = 0;
-    u32 saved_stream_id = 0;
-    int saved_buf_pos = 0;
-    u8 found_data_frame = 0;
-    http2_conn_stream_t stream = {0};
-    stream.pid_conn = *pid_conn;
-
-    unsigned char frame_buf[FRAME_HEADER_LEN];
-    frame_header_t frame = {0};
-
-    for (int i = 0; i < 8; i++) {
-        if (pos >= bytes_len) {
+static __always_inline frame_header_t next_frame(const grpc_frames_ctx_t *g_ctx) {
+    const call_protocol_args_t *args = &g_ctx->args;
+
+    frame_header_t header;
+
+    // read next frame
+    const void *offset = (u8 *)args->u_buf + g_ctx->pos;
+
+    if (bpf_probe_read(&header, sizeof(header), offset) != 0) {
+        bpf_dbg_printk("failed to read frame header");
+        goto header_error;
+    }
+
+    bpf_dbg_printk("http2 frame type = %d, len = %d", header.type, header.length);
+    bpf_dbg_printk("http2 frame stream_id = %d, flags = %d", header.stream_id, header.flags);
+
+    if (header.length || header.type > FrameContinuation) {
+        goto header_error;
+    }
+
+    header.length = bpf_ntohl(header.length << 8);
+    header.stream_id = bpf_ntohl(header.stream_id << 1);
+
+    return header;
+
+header_error:
+    header.type = FrameUnknown;
+    return header;
+}
+
+static __always_inline void update_prev_info(grpc_frames_ctx_t *g_ctx) {
+    if (g_ctx->has_prev_info) {
+        return;
+    }
+
+    const http2_grpc_request_t *prev_info =
+        bpf_map_lookup_elem(&ongoing_http2_grpc, &g_ctx->stream);
+
+    if (prev_info) {
+        g_ctx->prev_info = *prev_info;
```

Comment:
I like this change, it's safer to make a copy of the prev info rather than to keep a pointer. With high transaction rates we can potentially pick up stale data.

Reply:
I was also having trouble keeping the pointer on a map. If I stored it as a `u64` and tried to cast it back and deference it later on, the verifier would complain about "forbidden memory deref" or something like that.